node/pkg/cni/token_watch.go

```diff
@@ -1,56 +1,158 @@
+const defaultRefreshFraction = 3
+
 var kubeconfigPath string = "/host/etc/cni/net.d/calico-kubeconfig"
 
-func Run() {
-	// Log to stdout.  this prevents our logs from being interpreted as errors by, for example,
-	// fluentd's default configuration.
-	logrus.SetOutput(os.Stdout)
+type TokenRefresher struct {
+	tokenSupported bool
+	tokenOnce      *sync.Once
+
+	namespace          string
+	serviceAccountName string
 
-	// Create a watcher for file changes.
-	watcher, err := fsnotify.NewWatcher()
+	tokenChan chan TokenUpdate
+}
+
+type TokenUpdate struct {
+	Token          string
+	ExpirationTime time.Time
+	Error          error
+}
+
+func NewTokenRefresher(namespace string, serviceAccountName string) *TokenRefresher {
+	return &TokenRefresher{
+		tokenSupported:     false,
+		tokenOnce:          &sync.Once{},
+		namespace:          namespace,
+		serviceAccountName: serviceAccountName,
+		tokenChan:          make(chan TokenUpdate),
+	}
+}
+
+func (t *TokenRefresher) UpdateToken() TokenUpdate {
+	cfg, err := rest.InClusterConfig()
+	if err != nil {
+		logrus.WithError(err).Error("Error generating kube config.")
+		return tokenUpdateError(err)
+	}
+	err = rest.LoadTLSFiles(cfg)
+	if err != nil {
+		logrus.WithError(err).Error("Error loading TLS files.")
+		return tokenUpdateError(err)
+	}
+	clientset, err := kubernetes.NewForConfig(cfg)
+	if err != nil {
+		logrus.WithError(err).Error("Failed to create clientset for CNI kubeconfig")
+		return tokenUpdateError(err)
+	}
+
+	validity := int64(defaultCNITokenValiditySeconds)
+	tr := &authenticationv1.TokenRequest{
+		Spec: authenticationv1.TokenRequestSpec{
+			Audiences:         []string{"kubernetes"},
+			ExpirationSeconds: &validity,
+		},
+	}
+
+	tokenRequest, err := clientset.CoreV1().ServiceAccounts(t.namespace).CreateToken(context.TODO(), t.serviceAccountName, tr, metav1.CreateOptions{})
+	if apierrors.IsNotFound(err) && !t.tokenRequestSupported(clientset) {
+		logrus.WithError(err).Error("Unable to create token for CNI kubeconfig as token request api is not supported")
+		return tokenUpdateError(errors.New("token request api not supported"))
+	}
 	if err != nil {
-		panic(err)
+		logrus.WithError(err).Error("Unable to create token for CNI kubeconfig")
+		return tokenUpdateError(err)
 	}
-	defer watcher.Close()
 
-	// Watch for changes to the serviceaccount directory. Rety if necessary.
+	return TokenUpdate{
+		Token:          tokenRequest.Status.Token,
+		ExpirationTime: tokenRequest.Status.ExpirationTimestamp.Time,
+	}
+}
+
+func (t *TokenRefresher) TokenChan() <-chan TokenUpdate {
+	return t.tokenChan
+}
+
+func (t *TokenRefresher) Run() {
+	var nextExpiration time.Time
 	for {
-		if err := watcher.Add(serviceaccountDirectory); err != nil {
-			// Error watching the file - retry
-			logrus.WithError(err).Error("Failed to watch Kubernetes serviceaccount files.")
-			time.Sleep(5 * time.Second)
-			continue
+		tu := t.UpdateToken()
+		if tu.Error == nil {
```

Comment:
Hm, the `TokenUpdate` type has an `Error` on it, which is right if we were to be attempting to send errors over the token channel. However, it seems we only ever send a TokenUpdate if the error is nil, making that field a little bit confusing. So I think we should probably:

- Adjust that struct so that it no longer has an Error
- Adjust UpdateToken() so that it returns `(TokenUpdate, error)`
- Update this code so that it handles an error returned from `UpdateToken` 

We probably also want to set a shorter retry time in this loop should we see an error, so that we don't have to wait quite as long in the case of a transient error.

Reply:
Refactored according to your suggestion. Indeed having the error in the struct, but only propagation successful attempts over the channel was a bit confusing. Now, it should be more clear.